java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/BinaryRelation.java

```diff
@@ -292,65 +242,88 @@ protected BinaryRelation symmetric() {
    * </ul>
    * @see RelationState
    */
-  protected RelationState implies(BinaryRelation relation) {
-    if (leftOp.equals(relation.leftOp) && rightOp.equals(relation.rightOp)) {
-      return relation.isImpliedBy(this);
-    } else if (leftOp.equals(relation.rightOp) && rightOp.equals(relation.leftOp)) {
-      return relation.symmetric().isImpliedBy(this);
+  private RelationState implies(BinaryRelation relation) {
+    if (this.equals(relation)) {
+      return RelationState.FULFILLED;
+    }
+    if (inverse().equals(relation)) {
+      return RelationState.UNFULFILLED;
+    }
+    if (hasSameOperandsAs(relation)) {
+      return RelationStateTable.solveRelation(kind, relation.kind);
     }
     return RelationState.UNDETERMINED;
   }
 
-  protected abstract RelationState isImpliedBy(BinaryRelation relation);
-
-  protected abstract RelationState impliesEqual();
-
-  protected abstract RelationState impliesNotEqual();
-
-  protected abstract RelationState impliesMethodEquals();
-
-  protected abstract RelationState impliesNotMethodEquals();
-
-  protected abstract RelationState impliesGreaterThan();
-
-  protected abstract RelationState impliesGreaterThanOrEqual();
-
-  protected abstract RelationState impliesLessThan();
-
-  protected abstract RelationState impliesLessThanOrEqual();
-
-  /**
-   * Returns a new relation resulting of the transitive combination of the receiver with the supplied relation.
-   * Schema aRb & bSc => aTc
-   * where aRb is the relation described by the receiver, bSc that of the supplied relation, aTc is the returned relation.
-   * 
-   * @param relation another relation whose left operand is the same as the receiver's right operand
-   * @return the resulting relation or null if the conjunction of both relations does not bring any new information
-   */
-  @CheckForNull
-  protected abstract BinaryRelation combinedAfter(BinaryRelation relation);
-
-  @CheckForNull
-  protected abstract BinaryRelation combinedWithEqual(EqualRelation relation);
+  private static Stream<BinaryRelation> combineTransitively(BinaryRelation rel1, BinaryRelation rel2) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this method 7, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1142)